### ISSUE TITLE
cryptsetup status fails

### ISSUE THREAD

**Issue by mmcdole** [ejmg/an-idiots-guide-to-installing-arch-on-a-lenovo-carbon-x1-gen-6#10]
This is very likely an issue with cryptsetup / luks2, but I wanted to track an issue here just in case it ended up being a configuration issue.

When I attempt to run 'sudo cryptsetup status' on our encrypted partitions, I get the following error.

```
sudo cryptsetup status -v /dev/mapper/cryptgrp-home              
Command failed with code -1 (wrong or missing parameters).
```

Instead, I would expect output like the following:
```
/dev/mapper/cryptgrp-home is active.
  type:    LUKS1
  cipher:  aes-cbc-essiv:sha256
  keysize: 256 bits
  device:  /dev/xvdc
  offset:  4096 sectors
  size:    419426304 sectors
  mode:    read/write
Command successful.
```

Still investigating this issue!

**Comment by mmcdole** [comment]
Created the following issue: https://gitlab.com/cryptsetup/cryptsetup/issues/413

**Comment by ejmg** [comment]
I can report similar behavior on my install, which is disheartening. This especially makes me worried that if the luks encryption logic itself is now faulty/fialed, a full re-install with whatever fix would be required to actually benefit from LUKS running on lvm without going through some truly painful backup magic and reformatting magic.

That said, I think I have the answer for what is going wrong here and I now believe nothing is wrong.

# TL;DR

I do not make this explicit (indeed, my off-the-cuff notes may actually muddle it), but the encryption scheme I use in the tutorial is [LVM on LUKS](https://wiki.archlinux.org/index.php/Dm-crypt/Encrypting_an_entire_system#LVM_on_LUKS). TL;DR, we encrypt one big partition with LUKS first and then, after unlocking the partition, we break up the partition into LVM volumes proper. This makes the upfront logic simple but makes extending partitions in the future more or less impossible since they are all within this partition.

What this means is that the root and home LVM volumes **themselves** are not encrypted. The LUKS "container" they sit inside of, however, is encrypted. Come boot time, we un-encrypt `nvme0n1p2` which then gives us access to our lvm volumes `main_group-root` and so forth.

# Explanation

At face, it looks like something is wrong for sure. In addition to `cryptsetup status`, digging deeper into why `main_group-root` is not returning a successful message would make it appear that it is outright not encrypted at all.

This is what I get when I run `sudo cryptsetup --verbose --debug luksDump /dev/mapper/main_group-root`:

```
# cryptsetup 2.0.4 processing "cryptsetup --verbose --debug luksDump /dev/mapper/main_group-root"
# Running command luksDump.
# Locking memory.
# Installing SIGINT/SIGTERM handler.
# Unblocking interruption on signal.
# Allocating context for crypt device /dev/mapper/main_group-root.
# Trying to open and read device /dev/mapper/main_group-root with direct-io.
# Initialising device-mapper backend library.
# Trying to load any crypt type from device /dev/mapper/main_group-root.
# Crypto backend (gcrypt 1.8.3) initialized in cryptsetup library version 2.0.4.
# Detected kernel Linux 4.18.5-arch1-1-ARCH x86_64.
# Loading LUKS2 header (repair disabled).
# Opening lock resource file /run/cryptsetup/L_254:2
# Acquiring read lock for device /dev/mapper/main_group-root.
# Verifying read lock handle for device /dev/mapper/main_group-root.
# Device /dev/mapper/main_group-root READ lock taken.
# Trying to read primary LUKS2 header at offset 0x0.
# Opening locked device /dev/mapper/main_group-root
# Veryfing locked device handle (bdev)
# Trying to read secondary LUKS2 header at offset 0x4000.
# Opening locked device /dev/mapper/main_group-root
# Veryfing locked device handle (bdev)
# Trying to read secondary LUKS2 header at offset 0x8000.
# Opening locked device /dev/mapper/main_group-root
# Veryfing locked device handle (bdev)
# Trying to read secondary LUKS2 header at offset 0x10000.
# Opening locked device /dev/mapper/main_group-root
# Veryfing locked device handle (bdev)
# Trying to read secondary LUKS2 header at offset 0x20000.
# Opening locked device /dev/mapper/main_group-root
# Veryfing locked device handle (bdev)
# Trying to read secondary LUKS2 header at offset 0x40000.
# Opening locked device /dev/mapper/main_group-root
# Veryfing locked device handle (bdev)
# Trying to read secondary LUKS2 header at offset 0x80000.
# Opening locked device /dev/mapper/main_group-root
# Veryfing locked device handle (bdev)
# Trying to read secondary LUKS2 header at offset 0x100000.
# Opening locked device /dev/mapper/main_group-root
# Veryfing locked device handle (bdev)
# Trying to read secondary LUKS2 header at offset 0x200000.
# Opening locked device /dev/mapper/main_group-root
# Veryfing locked device handle (bdev)
# Trying to read secondary LUKS2 header at offset 0x400000.
# Opening locked device /dev/mapper/main_group-root
# Veryfing locked device handle (bdev)
# LUKS2 header read failed (-22).
# Device /dev/mapper/main_group-root READ lock released.
# Releasing crypt device /dev/mapper/main_group-root context.
# Releasing device-mapper backend.
# Unlocking memory.
Command failed with code -1 (wrong or missing parameters).
```
We fail and get the same error code.

This is because  cryptsetup is looking for a luks header but cannot find one: this would be because the volume itself is not encrypted.

*However*, if we look at the main partition itself that we encrypt in the tutorial with `sudo cryptsetup --verbose --debug luksDump /dev/nvme0n1p2`, we get something much more encouraging:

```
# cryptsetup 2.0.4 processing "cryptsetup --verbose --debug luksDump /dev/nvme0n1p2"
# Running command luksDump.
# Locking memory.
# Installing SIGINT/SIGTERM handler.
# Unblocking interruption on signal.
# Allocating context for crypt device /dev/nvme0n1p2.
# Trying to open and read device /dev/nvme0n1p2 with direct-io.
# Initialising device-mapper backend library.
# Trying to load any crypt type from device /dev/nvme0n1p2.
# Crypto backend (gcrypt 1.8.3) initialized in cryptsetup library version 2.0.4.
# Detected kernel Linux 4.18.5-arch1-1-ARCH x86_64.
# PBKDF pbkdf2, hash sha256, time_ms 2000 (iterations 0), max_memory_kb 0, parallel_threads 0.
# Reading LUKS header of size 1024 from device /dev/nvme0n1p2
# Key length 32, device size 999164559 sectors, header size 2050 sectors.
LUKS header information for /dev/nvme0n1p2

Version:        1
Cipher name:    aes
Cipher mode:    xts-plain64
Hash spec:      sha256
Payload offset: 4096
MK bits:        256
MK digest:      [RETRACTED]
MK salt:        [RETRACTED]
MK iterations:  [RETRACTED]
UUID:           93c0a958-43f5-4e3a-94ed-007881c2a098

Key Slot 0: ENABLED
        Iterations:             [RETRACTED]
        Salt:                  [RETRACTED]
        Key material offset:    [RETRACTED]
        AF stripes:             [RETRACTED]
Key Slot 1: DISABLED
Key Slot 2: DISABLED
Key Slot 3: DISABLED
Key Slot 4: DISABLED
Key Slot 5: DISABLED
Key Slot 6: DISABLED
Key Slot 7: DISABLED
# Releasing crypt device /dev/nvme0n1p2 context.
# Releasing device-mapper backend.
# Unlocking memory.
Command successful.
```
:tada:  **Command successful!** :tada: 

Furthermore, running the original command in question but with `cryptlvm` as our device, we get:

```
/dev/mapper/cryptlvm is active and is in use.
  type:    LUKS1
  cipher:  aes-xts-plain64
  keysize: 256 bits
  key location: dm-crypt
  device:  /dev/nvme0n1p2
  sector size:  512
  offset:  4096 sectors
  size:    999160463 sectors
  mode:    read/write
Command successful.
```

:tada: 

If you are wondering where `cryptlvm` came from, it is what we named our `cryptdevice` in our boot loader's `options` parameters in `/boot/loader/entries/arch.conf`. We added a label to it, `:cryptlvm`, which is what identifies it to the bootloader. This is what we need to check to ensure LUKS encryption was done successfully because it is the only thing actually encrypted. The abstraction of LVM is done within this partition and none of the volumes themselves are encrypted. At this point, I am beating a dead horse but I myself just had to run through a lot of wiki resources and docs to make sure there was nothing seriously wrong here.

# summary

i am 99% certain nothing is wrong after investigating the issue. It was a while since I was emerged in the technical aspects of this guide and had a momentary panic, but after revisiting the arch wikis and investigating how cryptsetup actually works, I realize my missteps in communication. Maybe we should put the `cryptsetup status cryptlvm` check in the guide to save other people from checking the wrong device for its encryption status?

**Comment by ejmg** [comment]
I'm going to close this issue for now, but please do let me know if what I have written/explained above assures you of the status of the encryption scheme used by the guide. Furthermore, I think clarifying this in the guide is a very good idea given the small technical details involved in a LVM on LUKS encryption scheme can easily be conflated with a LUKS on LVM encryption scheme and lead to other readers/users to throw a red flag.

Finally, thank you for double checking the encryption scheme itself. I am an absolute idiot and never once bothered to use these commands to verify the status of my partition. Only because you brought this up am I now aware of how to quickly check my cryptsetup status and navigate that as an issue.

**Comment by mmcdole** [comment]
Ah, shit. That is embarrassing.

Makes total sense that those logical volumes wouldn't show anything via cryptsetup status!  I even understood the distinction between "LVM on Luks" vs "Luks on LVM" and I still somehow made this mistake.

Anyway, just glad I get my 'warm and fuzzies' now that I can see the output on `cryptlvm`.  Thanks for the detailed reply!
